drivers/misc/cldatadrv/rcar_cluster_drv.c

```diff
@@ -0,0 +1,330 @@
+		goto end;
+	}
+	else {
+		memcpy(res_msg, event->result, sizeof(taurus_cluster_res_msg_t));
+	}
+
+	write_lock(&clusterdrv->event_list_lock);
+	list_del(&event->list);
+	write_unlock(&clusterdrv->event_list_lock);
+
+end:
+	devm_kfree(&rpdev->dev, event->result);
+	devm_kfree(&rpdev->dev, event);	
+
+	return ret;
+}
+
+/* -----------------------------------------------------------------------------
+ * RPMSG operations
+ */
+
+static int rpmsg_cluster_cb(struct rpmsg_device* rpdev, void* data, int len,
+			void* priv, u32 src) {
+	int ret = 0;
+	struct taurus_event_list* event = NULL;
+	struct taurus_cluster_res_msg* res = (struct taurus_cluster_res_msg*)data;
+	struct list_head* i = NULL;
+	rcar_cluster_device_t* clusterdrv = (rcar_cluster_device_t*)dev_get_drvdata(&rpdev->dev);
+	uint32_t res_id = res->hdr.Id;
+
+	if (!(res->hdr.Result == R_TAURUS_CMD_NOP && res_id == 0)) {/*necessary send data to cluster*/
+		/*send ACK message*/
+		read_lock(&clusterdrv->event_list_lock);
+		
+		list_for_each_prev(i, &clusterdrv->taurus_event_list_head) {
+			event = list_entry(i, struct taurus_event_list, list);
+			if (event->id == res_id) {
+				memcpy(event->result, data, len);
+				if (event->ack_received) {
+					dev_info(&rpdev->dev, "send_msg: Message completed (%d)\n", ret);
+					complete(&event->completed);
+				} else {
+					event->ack_received = 1;
+					complete(&event->ack);
+				}
+			}
+		}
+		read_unlock(&clusterdrv->event_list_lock);
+	}
+	return 0;
+}
+
+static int rpmsg_cluster_probe(struct rpmsg_device* rpdev)
+{
+	rcar_cluster_device_t *clusterdvc = NULL;
+	int ret = 0;
+	struct miscdevice *dev = NULL;
+	taurus_cluster_res_msg_t res_msg;
+
+	taurus_cluster_data_t cluster_data = {
+		.value = 10,
+		.ioctl_cmd = 1,
+	};
+	dev_info(&rpdev->dev, "cluster: send_msg: probe\n");
+
+	clusterdvc = kzalloc(sizeof(*clusterdvc), GFP_KERNEL);
+
+	if (clusterdvc== NULL)
+		return -ENOMEM;
+	
+	clusterdvc->rpdev = rpdev;
+	dev = &clusterdvc->dev;
+	dev->parent = &rpdev->dev;
+	dev->minor = MISC_DYNAMIC_MINOR;
+
+	mutex_lock(&idr_lock);
+	ret = idr_alloc(&misc_idr, clusterdvc, 0, MAX_MISC_RPMSG, GFP_KERNEL);
+	mutex_unlock(&idr_lock);
+
+	if (ret < 0) {
+		goto free_cluster_data;
+	}
+
+	dev->name = kstrdup("cluster-taurus", GFP_KERNEL);
+	if (!dev->name) {
+		ret = -ENOMEM;
+		goto free_name;
+	}
+
+	dev->fops = &misc_fops;
+	ret = misc_register(dev);
+	
+	dev_set_drvdata(&rpdev->dev, clusterdvc);
+	
+	INIT_LIST_HEAD(&clusterdvc->taurus_event_list_head);
+	rwlock_init(&clusterdvc->event_list_lock);
+	/*send a ping of message with dummy data*/
+	send_msg(rpdev, &cluster_data, &res_msg);
+
+	return ret;
+
+free_name:
+	kfree(dev->name);
+	dev->name = NULL;
+
+free_cluster_data:
+	kfree(clusterdvc);
+
+	return ret;
+}
+
+static void rpmsg_cluster_remove(struct rpmsg_device *rpdev)
+{
+	rcar_cluster_device_t *data = dev_get_drvdata(&rpdev->dev);
+	
+	mutex_lock(&idr_lock);
+	idr_remove(&misc_idr, 0);
```

Comment:
And there you use hardcoded IDR `0`.

I believe, you can just remove all this stuff, if you are not using it properly